### PR TITLE
Fix race condition during CRMP sendMessage

### LINKED ISSUE #6155: [Cirque]chip-im-initiat" received signal SIGSEGV, Segmentation fault.
#### Problem

Thread 8 "chip-im-initiat" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fffe77fe700 (LWP 169)]
0x000055555558d434 in chip::System::PacketBuffer::DataLength (this=0x0)
    at ../../../../../src/system/SystemPacketBuffer.h:217
#0  0x000055555558d434 in chip::System::PacketBuffer::DataLength() const
    (this=0x0) at ../../../../../src/system/SystemPacketBuffer.h:217
#1  0x00005555555bf5ae in chip::EncryptedPacketBufferHandle::GetMsgId() const
    (this=0x555555616570)
    at ../../../../../src/transport/SecureSessionMgr.cpp:61
#2  0x000055555559bf20 in chip::Messaging::ReliableMessageMgr::CheckAndRemRetransTable(chip::Messaging::ReliableMessageContext*, unsigned int)
    (this=0x555555616538, rc=0x555555616720, ackMsgId=0)
    at ../../../../../src/messaging/ReliableMessageMgr.cpp:327
#3  0x000055555559b190 in chip::Messaging::ReliableMessageContext::HandleRcvdAck(unsigned int) (this=0x555555616720, AckMsgId=0)
    at ../../../../../src/messaging/ReliableMessageContext.cpp:166
#4  0x00005555555979fb in chip::Messaging::ExchangeMessageDispatch::OnMessageReceived(chip::PayloadHeader const&, unsigned int, chip::Transport::PeerAddress const&, chip::Messaging::ReliableMessageContext&)
    (this=0x5555555d6418 <chip::app::sInteractionModelEngine+600>, payloadHeader=..., messageId=0, peerAddress=..., reliableMessageContext=...)
    at ../../../../../src/messaging/ExchangeMessageDispatch.cpp:112
#5  0x0000555555596f28 in chip::Messaging::ExchangeContext::HandleMessage(chip::PacketHeader const&, chip::PayloadHeader const&, chip::Transport::PeerAddress const&, chip::System::PacketBufferHandle)
    (this=0x555555616718, packetHeader=..., payloadHeader=..., peerAddress=..., msgBuf=...) at ../../../../../src/messaging/ExchangeContext.cpp:380
#6  0x00005555555987fb in chip::Messaging::ExchangeManager::OnMessageReceived(chip::PacketHeader const&, chip::PayloadHeader const&, chip::SecureSessionHandle, chip::Transport::PeerAddress const&, chip::System::PacketBufferHandle, chip::SecureSessionMgr*)
    (this=0x555555616510, packetHeader=..., payloadHeader=..., session=..., source=..., msgBuf=..., msgLayer=0x555555617640)
    at ../../../../../src/messaging/ExchangeMgr.cpp:234
#7  0x00005555555c0e83 in chip::SecureSessionMgr::SecureMessageDispatch(chip::PacketHeader const&, chip::Transport::PeerAddress const&, chip::System::PacketBufferHandle) (this=0x555555617640, packetHeader=..., peerAddress=..., msg=...)
    at ../../../../../src/transport/SecureSessionMgr.cpp:465
#8  0x00005555555c0830 in chip::SecureSessionMgr::OnMessageReceived(chip::PacketHeader const&, chip::Transport::PeerAddress const&, chip::System::PacketBufferHandle) (this=0x555555617640, packetHeader=..., peerAddress=..., msg=...)
    at ../../../../../src/transport/SecureSessionMgr.cpp:356
#9  0x00005555555c19cc in chip::TransportMgrBase::HandleMessageReceived(chip::PacketHeader const&, chip::Transport::PeerAddress const&, chip::System::PacketBufferHandle) (this=0x5555556057b0, packetHeader=..., peerAddress=..., msg=...)
    at ../../../../../src/transport/TransportMgrBase.cpp:59
#10 0x00005555555c3021 in chip::Transport::Base::HandleMessageReceived(chip::PacketHeader const&, chip::Transport::PeerAddress const&, chip::System::PacketBufferHandle&&) (this=0x5555556057d8, header=..., source=..., buffer=...)
    at ../../../../../src/transport/raw/Base.h:93
#11 0x00005555555c356c in chip::Transport::UDP::OnUdpReceive(chip::Inet::IPEndPointBasis*, chip::System::PacketBufferHandle, chip::Inet::IPPacketInfo const*)
    (endPoint=0x5555555d89e0 <chip::Inet::UDPEndPoint::sPool>, buffer=..., pktInfo=0x7fffe77fdab0) at ../../../../../src/transport/raw/UDP.cpp:118
#12 0x00005555555a8354 in chip::Inet::IPEndPointBasis::HandlePendingIO(unsigned short) (this=0x5555555d89e0 <chip::Inet::UDPEndPoint::sPool>, aPort=11098)
    at ../../../../../src/inet/IPEndPointBasis.cpp:1202
#13 0x00005555555a6a91 in chip::Inet::UDPEndPoint::HandlePendingIO()
    (this=0x5555555d89e0 <chip::Inet::UDPEndPoint::sPool>)
    at ../../../../../src/inet/UDPEndPoint.cpp:899
#14 0x00005555555a44c9 in chip::Inet::InetLayer::HandleSelectResult(int, fd_set*, fd_set*, fd_set*) (this=
    0x5555555d6760 <chip::DeviceLayer::InetLayer>, selectRes=11, readfds=0x5555555d6810 <chip::DeviceLayer::PlatformManagerImpl::sInstance+16>, writefds=0x5555555d6890 <chip::DeviceLayer::PlatformManagerImpl::sInstance+144>, exceptfds=0x5555555d6910 <chip::DeviceLayer::PlatformManagerImpl::sInstance+272>)
    at ../../../../../src/inet/InetLayer.cpp:1234
#15 0x00005555555775fa in chip::DeviceLayer::Internal::GenericPlatformManagerImpl_POSIX<chip::DeviceLayer::PlatformManagerImpl>::SysProcess()
    (this=0x5555555d6800 <chip::DeviceLayer::PlatformManagerImpl::sInstance>)
    at ../../../../../src/include/platform/internal/GenericPlatformManagerImpl_POSIX.cpp:185
#16 0x000055555557718e in chip::DeviceLayer::Internal::GenericPlatformManagerImpl_POSIX<chip::DeviceLayer::PlatformManagerImpl>::_RunEventLoop()
    (this=0x5555555d6800 <chip::DeviceLayer::PlatformManagerImpl::sInstance>)
    at ../../../../../src/include/platform/internal/GenericPlatformManagerImpl_POSIX.cpp:203
#17 0x0000555555576b0c in chip::DeviceLayer::PlatformManager::RunEventLoop()
    (this=0x5555555d6800 <chip::DeviceLayer::PlatformManagerImpl::sInstance>)
    at ../../../../../src/include/platform/PlatformManager.h:204
#18 0x0000555555577705 in chip::DeviceLayer::Internal::GenericPlatformManagerImpl_POSIX<chip::DeviceLayer::PlatformManagerImpl>::EventLoopTaskMain(void*)
    (arg=0x5555555d6800 <chip::DeviceLayer::PlatformManagerImpl::sInstance>)
    at ../../../../../src/include/platform/internal/GenericPlatformManagerImpl_POSIX.cpp:214
#19 0x00007ffff7fab609 in start_thread ()
    at /lib/x86_64-linux-gnu/libpthread.so.0
#20 0x00007ffff763d293 in clone () at /lib/x86_64-linux-gnu/libc.so.6
(gdb) quit
A debugging session is active.
 
	Inferior 1 [process 159] will be killed.
 
Quit anyway? (y or n) [answered Y; input not from terminal]
 
2021-04-19 02:05:11,469 [CHIPInteractionModelTest] ERROR IM failure: Reading symbols from chip-im-initiator...
Starting program: /usr/bin/chip-im-initiator 172.20.0.3
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7ffff6f29700 (LWP 163)]
[New Thread 0x7ffff6728700 (LWP 164)]
[New Thread 0x7ffff5f27700 (LWP 165)]
[New Thread 0x7ffff5726700 (LWP 166)]
Init CHIP Stack
CHIP:DL: writing settings to file (/tmp/chip_factory.ini.tmp)
CHIP:DL: renamed tmp file to file (/tmp/chip_factory.ini)
CHIP:DL: writing settings to file (/tmp/chip_config.ini.tmp)
CHIP:DL: renamed tmp file to file (/tmp/chip_config.ini)
CHIP:DL: writing settings to file (/tmp/chip_counters.ini.tmp)
CHIP:DL: renamed tmp file to file (/tmp/chip_counters.ini)
[New Thread 0x7ffff4f25700 (LWP 167)]
[New Thread 0x7fffe7fff700 (LWP 168)]
CHIP:DL: CHIP task running
CHIP:IN: TransportMgr initialized
CHIP:IN: local node id is 0x000000000001b669
CHIP:IN: New pairing for device 0x0000000000bc5c01, key 0!!
[New Thread 0x7fffe77fe700 (LWP 169)]
Establish secure session succeeded
CHIP:DMG: ICR moving to [Initialize]
 
Send invoke command request message to Node: 12344321
CHIP:DMG: ICR moving to [AddCommand]
CHIP:IN: Secure message was encrypted: Msg ID 0
CHIP:IN: Sending msg from 0x000000000001b669 to 0x0000000000bc5c01 at utc time: 734991 msec
CHIP:IN: Sending secure msg on generic transport
CHIP:IN: Secure msg send status No Error
CHIP:DMG: ICR moving to [   Sending]
217	../../../../../src/system/SystemPacketBuffer.h: No such file or directory.
 
Thread 8 "chip-im-initiat" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fffe77fe700 (LWP 169)]
0x000055555558d434 in chip::System::PacketBuffer::DataLength (this=0x0)
    at ../../../../../src/system/SystemPacketBuffer.h:217
#0  0x000055555558d434 in chip::System::PacketBuffer::DataLength() const
    (this=0x0) at ../../../../../src/system/SystemPacketBuffer.h:217
#1  0x00005555555bf5ae in chip::EncryptedPacketBufferHandle::GetMsgId() const
    (this=0x555555616570)
    at ../../../../../src/transport/SecureSessionMgr.cpp:61
#2  0x000055555559bf20 in chip::Messaging::ReliableMessageMgr::CheckAndRemRetransTable(chip::Messaging::ReliableMessageContext*, unsigned int)
    (this=0x555555616538, rc=0x555555616720, ackMsgId=0)
    at ../../../../../src/messaging/ReliableMessageMgr.cpp:327
#3  0x000055555559b190 in chip::Messaging::ReliableMessageContext::HandleRcvdAck(unsigned int) (this=0x555555616720, AckMsgId=0)
    at ../../../../../src/messaging/ReliableMessageContext.cpp:166
#4  0x00005555555979fb in chip::Messaging::ExchangeMessageDispatch::OnMessageReceived(chip::PayloadHeader const&, unsigned int, chip::Transport::PeerAddress const&, chip::Messaging::ReliableMessageContext&)
    (this=0x5555555d6418 <chip::app::sInteractionModelEngine+600>, payloadHeader=..., messageId=0, peerAddress=..., reliableMessageContext=...)
    at ../../../../../src/messaging/ExchangeMessageDispatch.cpp:112
#5  0x0000555555596f28 in chip::Messaging::ExchangeContext::HandleMessage(chip::PacketHeader const&, chip::PayloadHeader const&, chip::Transport::PeerAddress const&, chip::System::PacketBufferHandle)
    (this=0x555555616718, packetHeader=..., payloadHeader=..., peerAddress=..., msgBuf=...) at ../../../../../src/messaging/ExchangeContext.cpp:380
#6  0x00005555555987fb in chip::Messaging::ExchangeManager::OnMessageReceived(chip::PacketHeader const&, chip::PayloadHeader const&, chip::SecureSessionHandle, chip::Transport::PeerAddress const&, chip::System::PacketBufferHandle, chip::SecureSessionMgr*)
    (this=0x555555616510, packetHeader=..., payloadHeader=..., session=..., source=..., msgBuf=..., msgLayer=0x555555617640)
    at ../../../../../src/messaging/ExchangeMgr.cpp:234
#7  0x00005555555c0e83 in chip::SecureSessionMgr::SecureMessageDispatch(chip::PacketHeader const&, chip::Transport::PeerAddress const&, chip::System::PacketBufferHandle) (this=0x555555617640, packetHeader=..., peerAddress=..., msg=...)
    at ../../../../../src/transport/SecureSessionMgr.cpp:465
#8  0x00005555555c0830 in chip::SecureSessionMgr::OnMessageReceived(chip::PacketHeader const&, chip::Transport::PeerAddress const&, chip::System::PacketBufferHandle) (this=0x555555617640, packetHeader=..., peerAddress=..., msg=...)
    at ../../../../../src/transport/SecureSessionMgr.cpp:356
#9  0x00005555555c19cc in chip::TransportMgrBase::HandleMessageReceived(chip::PacketHeader const&, chip::Transport::PeerAddress const&, chip::System::PacketBufferHandle) (this=0x5555556057b0, packetHeader=..., peerAddress=..., msg=...)
    at ../../../../../src/transport/TransportMgrBase.cpp:59
#10 0x00005555555c3021 in chip::Transport::Base::HandleMessageReceived(chip::PacketHeader const&, chip::Transport::PeerAddress const&, chip::System::PacketBufferHandle&&) (this=0x5555556057d8, header=..., source=..., buffer=...)
    at ../../../../../src/transport/raw/Base.h:93
#11 0x00005555555c356c in chip::Transport::UDP::OnUdpReceive(chip::Inet::IPEndPointBasis*, chip::System::PacketBufferHandle, chip::Inet::IPPacketInfo const*)
    (endPoint=0x5555555d89e0 <chip::Inet::UDPEndPoint::sPool>, buffer=..., pktInfo=0x7fffe77fdab0) at ../../../../../src/transport/raw/UDP.cpp:118
#12 0x00005555555a8354 in chip::Inet::IPEndPointBasis::HandlePendingIO(unsigned short) (this=0x5555555d89e0 <chip::Inet::UDPEndPoint::sPool>, aPort=11098)
    at ../../../../../src/inet/IPEndPointBasis.cpp:1202
#13 0x00005555555a6a91 in chip::Inet::UDPEndPoint::HandlePendingIO()
    (this=0x5555555d89e0 <chip::Inet::UDPEndPoint::sPool>)
    at ../../../../../src/inet/UDPEndPoint.cpp:899
#14 0x00005555555a44c9 in chip::Inet::InetLayer::HandleSelectResult(int, fd_set*, fd_set*, fd_set*) (this=
    0x5555555d6760 <chip::DeviceLayer::InetLayer>, selectRes=11, readfds=0x5555555d6810 <chip::DeviceLayer::PlatformManagerImpl::sInstance+16>, writefds=0x5555555d6890 <chip::DeviceLayer::PlatformManagerImpl::sInstance+144>, exceptfds=0x5555555d6910 <chip::DeviceLayer::PlatformManagerImpl::sInstance+272>)
    at ../../../../../src/inet/InetLayer.cpp:1234
#15 0x00005555555775fa in chip::DeviceLayer::Internal::GenericPlatformManagerImpl_POSIX<chip::DeviceLayer::PlatformManagerImpl>::SysProcess()
    (this=0x5555555d6800 <chip::DeviceLayer::PlatformManagerImpl::sInstance>)
    at ../../../../../src/include/platform/internal/GenericPlatformManagerImpl_POSIX.cpp:185
#16 0x000055555557718e in chip::DeviceLayer::Internal::GenericPlatformManagerImpl_POSIX<chip::DeviceLayer::PlatformManagerImpl>::_RunEventLoop()
    (this=0x5555555d6800 <chip::DeviceLayer::PlatformManagerImpl::sInstance>)
    at ../../../../../src/include/platform/internal/GenericPlatformManagerImpl_POSIX.cpp:203
#17 0x0000555555576b0c in chip::DeviceLayer::PlatformManager::RunEventLoop()
    (this=0x5555555d6800 <chip::DeviceLayer::PlatformManagerImpl::sInstance>)
    at ../../../../../src/include/platform/PlatformManager.h:204
#18 0x0000555555577705 in chip::DeviceLayer::Internal::GenericPlatformManagerImpl_POSIX<chip::DeviceLayer::PlatformManagerImpl>::EventLoopTaskMain(void*)
    (arg=0x5555555d6800 <chip::DeviceLayer::PlatformManagerImpl::sInstance>)
    at ../../../../../src/include/platform/internal/GenericPlatformManagerImpl_POSIX.cpp:214
#19 0x00007ffff7fab609 in start_thread ()
    at /lib/x86_64-linux-gnu/libpthread.so.0
#20 0x00007ffff763d293 in clone () at /lib/x86_64-linux-gnu/libc.so.6
(gdb) quit
A debugging session is active.
 
	Inferior 1 [process 159] will be killed.
 
Quit anyway? (y or n) [answered Y; input not from terminal]
 
Traceback (most recent call last):
  File "/home/runner/work/connectedhomeip/connectedhomeip/src/test_driver/linux-cirque/helper/CHIPTestBase.py", line 62, in run_test
    self.test_routine()
  File "/home/runner/work/connectedhomeip/connectedhomeip/src/test_driver/linux-cirque/test-interaction-model.py", line 64, in test_routine
    self.run_data_model_test()
  File "/home/runner/work/connectedhomeip/connectedhomeip/src/test_driver/linux-cirque/test-interaction-model.py", line 79, in run_data_model_test
    self.assertEqual(
  File "/home/runner/work/connectedhomeip/connectedhomeip/src/test_driver/linux-cirque/helper/CHIPTestBase.py", line 222, in assertEqual
    raise AssertionError
AssertionError
2021-04-19 02:05:11,494 [CHIPInteractionModelTest] INFO destroying home: 8a3c78a9-df0f-4727-9864-8f0f65342dc5
Cleanup Flask
scripts/tests/cirque_tests.sh: line 58: kill: (7281) - No such process
scripts/tests/cirque_tests.sh: line 58: kill: (7286) - No such process
Do docker container and network prune
802 777 0:26 /docker.sock /run/docker.sock rw,nosuid,nodev - tmpfs tmpfs rw,size=712124k,mode=755
Test log can be found at /tmp/cirque_test_output/InteractionModelTest/device_logs

 #### Proposed Solution
<suggested fix, suggested enhancement>

### PR DESCRIPTION
**Pull request by yufengwangca**
#### Problem
Currently, we see the cirque is flaky with random crash if we enable IM and Echo test. The crash log shows we have race condition during CRMP sendMessage.

Now we need to first send the message and cache the message during CRMP since the packetHeader is encoded at TransportLayer during send. If we receive ack response before we cache the sent message, we will try to remove an non-exist buffer from retransTable and crash.

With #6333, packetHeader encoding and decoding is moved to SecureSessionMgr, we can cache the outgoing message first and send later to prevent this race condition.

 #### Summary of Changes
1. Fix race condition during CRMP sendMessage 
2. Enable EchoTest in Cirque
3. Enable InteractionModel Test in Cirque

 Fixes #6155

### CODE REVIEW

**Review by msandstedt — APPROVED** [review]

**Comment by pan-apple** [comment]
@yufengwangca is this PR dependent on #6333 to merge first?

**Comment by yufengwangca** [comment]
> @yufengwangca is this PR dependent on #6333 to merge first?

Yes, lets hold it until #6333 land

File: scripts/tests/cirque_tests.sh [hunk 1]
```diff
@@ -34,6 +34,7 @@ OT_SIMULATION_CACHE="$CIRQUE_CACHE_PATH/ot-simulation.tgz"
 
 # Append test name here to add more tests for run_all_tests
 CIRQUE_TESTS=(
+    "EchoTest"
```
Comment (erjiaqing) [inline scripts/tests/cirque_tests.sh:37]: Please try to add InteractionModel test as well.

Comment (yufengwangca) [reply scripts/tests/cirque_tests.sh:37]: ack

File: src/transport/SecureSessionMgr.cpp [hunk 2]
```diff
@@ -187,7 +187,8 @@ CHIP_ERROR SecureSessionMgr::SendMessage(SecureSessionHandle session, PayloadHea
     // Retain the packet buffer in case it's needed for retransmissions.
     if (bufferRetainSlot != nullptr)
     {
-        encryptedMsg = msgBuf.Retain();
+        encryptedMsg        = msgBuf.Retain();
```
Comment (andy31415) [inline src/transport/SecureSessionMgr.cpp:190]: why do we save the encryptedMsg first? Why not:

```cpp
*bufferRetainSlot = msgBuf.Retain();
```

that should make it clear that encryptedMsg does not contain a valid value here. We could make it even `encryptedMsg = nullptr` if we want to make it even more obvious.

**Review by andy31415 — CHANGES_REQUESTED** [review]
Could you describe the sequence of operations here? I have a hard time understanding the fix:

I believe our current code assumes single-threading (one main loop that handles all processing). In that case, the placement of where we save the retain slot does not seem to matter.

**Comment by andy31415** [comment]
Note that I agree that not trying to do the caching part in two sections of the code is a good change. I just am not clear how this fixes any form of race condition. 

Also 'encryptedMsg' should be deleted as a variable - it has no use.

**Comment by erjiaqing** [comment]
> Note that I agree that not trying to do the caching part in two sections of the code is a good change. I just am not clear how this fixes any form of race condition.
> 
> Also 'encryptedMsg' should be deleted as a variable - it has no use.

In very rare cases, the ack is received before SendMessage is returned, but the retrans table is not ready, thus it crashes.

This should not happen in real devices, but is somehow very frequent in github tests.

And, yes, we should do real send / receive works in one thread / task, all operations to CHIP objects should be guarded by mutex, but we need to figure the border of CHIP sdk first, this is tracked in https://github.com/project-chip/connectedhomeip/issues/6251

**Review by erjiaqing — APPROVED** [review]
Checked out this PR locally, seems cirque / darwin tests won't crash now.

**Comment by msandstedt** [comment]
> > Note that I agree that not trying to do the caching part in two sections of the code is a good change. I just am not clear how this fixes any form of race condition.
> > Also 'encryptedMsg' should be deleted as a variable - it has no use.
> 
> In very rare cases, the ack is received before SendMessage is returned, but the retrans table is not ready, thus it crashes.
> 
> This should not happen in real devices, but is somehow very frequent in github tests.
> 
> And, yes, we should do real send / receive works in one thread / task, all operations to CHIP objects should be guarded by mutex, but we need to figure the border of CHIP sdk first, this is tracked in #6251

I took it for granted that for this to occur in the first place, send must be happening from a different thread context than chip UDP comms.  This implies that send must be mutex-protected.

That mutex protection then also acts as a memory barrier: you cache first, lock, send, unlock.  This guarantees that caching will have occurred before unlock.

But this was just an assumption.  I was not able to trace the code to figure out where this locking would occur.  I assume it has to be within, or at least dependent upon platform-specific code.

TLDR: yes, the send operation needs to be mutex protected if send can ever occur from a different thread context.  Again, I assumed we had such protection.

File: src/transport/SecureSessionMgr.cpp (hunk 2 above)
Comment (yufengwangca) [reply src/transport/SecureSessionMgr.cpp]: The original purpose is to hold the reference to encrypted message in stack variable. In case of any failures, the reference is not returned, and this stack variable will automatically free the reference on returning from the function.

If we use *bufferRetainSlot to hold the retained buffer directly, we don't need encryptedMsg at all.

**Comment by yufengwangca** [comment]
> Could you describe the sequence of operations here? I have a hard time understanding the fix:
> 
> I believe our current code assumes single-threading (one main loop that handles all processing). In that case, the placement of where we save the retain slot does not seem to matter.

Currently, sendMessage is triggered from app within app's own thread directly, and receiveMessage is handled within CHIP  main thread. I think we should force a single-threading mode (This need certain refactor on the app side) in long term. That will avoid a lot of race issues.

**Comment by yufengwangca** [comment]
> > > Note that I agree that not trying to do the caching part in two sections of the code is a good change. I just am not clear how this fixes any form of race condition.
> > > Also 'encryptedMsg' should be deleted as a variable - it has no use.
> > 
> > 
> > In very rare cases, the ack is received before SendMessage is returned, but the retrans table is not ready, thus it crashes.
> > This should not happen in real devices, but is somehow very frequent in github tests.
> > And, yes, we should do real send / receive works in one thread / task, all operations to CHIP objects should be guarded by mutex, but we need to figure the border of CHIP sdk first, this is tracked in #6251
> 
> I took it for granted that for this to occur in the first place, send must be happening from a different thread context than chip UDP comms. This implies that send must be mutex-protected.
> 
> That mutex protection then also acts as a memory barrier: you cache first, lock, send, unlock. This guarantees that caching will have occurred before unlock.
> 
> But this was just an assumption. I was not able to trace the code to figure out where this locking would occur. I assume it has to be within, or at least dependent upon platform-specific code.
> 
> TLDR: yes, the send operation needs to be mutex protected if send can ever occur from a different thread context. Again, I assumed we had such protection.

The current CHIP messaging stack is largely inherited from Weave which has single-thread -model in mind, we only have one big lock PlatformManager::LockChipStack which is used to lock the whole chip stack to avoid contention from other application thread.

I agree we need mutex protection for sendMessage since it is triggered from app thread directly right now. Need to handle it in the separate PR and we still need to leverage if we should force a single thread model or use separate mutex.

**Comment by msandstedt** [comment]
> > I took it for granted that for this to occur in the first place, send must be happening from a different thread context than chip UDP comms. This implies that send must be mutex-protected.
> > That mutex protection then also acts as a memory barrier: you cache first, lock, send, unlock. This guarantees that caching will have occurred before unlock.
> > But this was just an assumption. I was not able to trace the code to figure out where this locking would occur. I assume it has to be within, or at least dependent upon platform-specific code.
> > TLDR: yes, the send operation needs to be mutex protected if send can ever occur from a different thread context. Again, I assumed we had such protection.
> 
> The current CHIP messaging stack is largely inherited from Weave which has single-thread -model in mind, we only have one big lock PlatformManager::LockChipStack which is used to lock the whole chip stack to avoid contention from other application thread.
> 
> I agree we need mutex protection for sendMessage since it is triggered from app thread directly right now. Need to handle it in the separate PR and we still need to leverage if we should force a single thread model or use separate mutex.

Thanks for considering my comments.

It's my personal belief that code shouldn't be written to assume a single-threaded environment and that code underlying public interfaces should protect itself to accommodate cross-thread calls.  But it's OK if the SDK isn't written this way.  We just need to know how to safely call into it.

Based on the above, is it safe to simply wrap cross-thread calls into the sdk with PlatformManager::Lock/UnlockChipStack?  That's an easy rule to follow.  However I'm having a tough time figuring out whether our current examples do this.  It appears things like chip-tool do not.

**Review by bzbarsky-apple — APPROVED** [review]
We really need to get our threading story straight, but if this unblocks us by preventing flakiness in CI, let's do it.  We just need to make sure we do the real fix of locking or serializing on a single thread appropriately as well...

**Review by andy31415 — APPROVED** [review]
Aproving - my guess is that this passes in a pointer directly in the retransmission table, hence the race.

However the real bug seems to be us using separate threads on a stack that strongly assumes single threading. This code is not yet safe - this is a temporary workaround rather than a fix.

**Comment by mrjerryjohns** [comment]
I've filed an issue here #6841 to get this sorted out. Once we do, can we please get this documented?
